### PR TITLE
Implement migrator that creates/populates `workflow_chain_set` (file: `migrator_from_X_to_PR9.py`)

### PR DESCRIPTION
**Pull request by brynnz22**
Migrator creates the `workflow_chain_set` and maps all `WorkflowExecution` subclasses to a workflow chain instance by the omics processing id in the `was_informed_by` slot.

- Migrator needs to come **Before** class name change from `OmicsProcessing` to `DataGeneration` as it refers to the `omics_processing_set`.

Address changes in PR #9

### CODE REVIEW

**Comment by eecavanna** [comment]
Here's the command I'm using to run the doctests in this migrator:

```shell
poetry run python -m doctest -v nmdc_schema/migrators/migrator_from_X_to_PR9.py
```

**Comment by eecavanna** [comment]
Hi @brynnz22, (a) I updated the doctest for the `populate_workflow_chain` method so that it passes, and (b) I updated all the doctests that were producing the "No adapter was specified. Migration capability will be limited" message so that they don't produce it anymore.

### Item (a)

For item (a), I created a "database" (a Python dictionary, since we do have an adapter for those), created an adapter _connected_ to that "database," then passed that adapter to the `Migrator` constructor. I also added an `import` statement, since `DictionaryAdapter` isn't actually used by the module (the module's _non-doctest_ code, that is).

```diff
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {}  # in this example, our data store is a Python dictionary
+        >>> adapter = DictionaryAdapter(database=database)
-        >>> m = Migrator()
+        >>> m = Migrator(adapter=adapter)
```

Then, at the end of the doctest, I used an adapter method to retrieve a document of interest.

```diff
-        # After execution, the workflow_chain_set collection should contain a document with the following data:
-        {'id': 'nmdc:wfc-456', ... 'name': 'Workflow Chain for metagenome analysis related to Study 1'}
+        >>> adapter.get_document_having_value_in_field('workflow_chain_set', 'id', value='nmdc:wfc-456')
+        {'id': 'nmdc:wfc-456', ... 'name': 'Workflow Chain for metagenome analysis of nmdc:omcp-123'}
```

I'll think about how I might be able to update the migration framework to simplify this kind of testing in the future.

### Item (b)

For item (b), I did the same thing as in the first part of what I did for item (a). I just wrote it in a more condensed way.

Here's an example:

```diff
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
-        >>> m = Migrator()
+        >>> m = Migrator(adapter=DictionaryAdapter(database={}))
```

I'm going to "un-assign" myself from this PR now to indicate more explicitly that I'm not planning to make any changes to the branch anymore. You can "re-assign" me again if you want me to make changes.

**Comment by brynnz22** [comment]
Thanks so much for looking into this and fixing it! I'll implement the same things you did in the instrument migration.

**Comment by eecavanna** [comment]
The logs from the most recent GitHub Actions failure says:

```
  RuntimeError

  Unable to find installation candidates for bioregistry (0.10.157)
```

The installation of a Python package failed. The failure is not related to the doctests.

**Comment by brynnz22** [comment]
> The log from the most recent GitHub Actions failure says:
> 
> ```
>   RuntimeError
> 
>   Unable to find installation candidates for bioregistry (0.10.157)
> ```
> 
> The installation of a Python package failed. The failure is not related to the doctests.
> 
> I opened a ticket in `nmdc-schema` ([microbiomedata#1932](https://github.com/microbiomedata/nmdc-schema/issues/1932)) about this failure. The failure is occurring in that repository also.

Katherine has a PR up to solve this, so its being worked on.

File: nmdc_schema/migrators/migrator_from_X_to_PR9.py
Comment (eecavanna) [inline nmdc_schema/migrators/migrator_from_X_to_PR9.py:233]: When I ran this migrator on a copy of the production database (a copy that had already undergone migration by the six migrators listed below), this line of code produced a `TypeError`.

image

I take this to mean `self.adapter.get_document_having_value_in_field()` returned `None`; which would happen if the adapter found no such document in the database. Maybe this is the behavior the author of this migrator expected while the `workflow_chain_ids.yaml` YAML file is empty (in which case, I'll hold off on testing this further).

Here are the six migrators that had already been run on the database in question:
```py
from nmdc_schema.migrators.migrator_from_X_to_PR23 import Migrator as M1
from nmdc_schema.migrators.migrator_from_X_to_PR4 import Migrator as M2
from nmdc_schema.migrators.migrator_from_X_to_PR53 import Migrator as M3
from nmdc_schema.migrators.migrator_from_X_to_PR21 import Migrator as M4
from nmdc_schema.migrators.migrator_from_X_to_PR129 import Migrator as M5
from nmdc_schema.migrators.migrator_from_X_to_PR31 import Migrator as M6
```

**Comment by eecavanna** [comment]
Implementing of this migrator is currently blocked by the Minter (or, rather, "a Minter") not allowing people to mint IDs for `WorkflowChain`s. I consider myself — @eecavanna — responsible for resolving that (with respect to this PR).

File: nmdc_schema/migrators/migrator_from_X_to_PR9.py
Comment (brynnz22) [reply nmdc_schema/migrators/migrator_from_X_to_PR9.py:233]: Oh yes sorry this migrator is broken for now cause I pushed changes to get ahead of the preminting process. I will finish it up today!

**Comment by eecavanna** [comment]
I'm reviewing this now. I will make some changes to variable names and push a new commit today.

**Comment by eecavanna** [comment]
As @brynnz22 said, the lines this PR shows as being _added_ to the `project.Makefile`, ([already exist](https://github.com/microbiomedata/berkeley-schema-fy24/blob/fe573e36ec139022a04fa5936921127bde7096e6/project.Makefile#L639-L671) in that file on the tip of the `main` branch. I don't understand why GitHub is showing them as new lines in this PR. 🤷

**Review by eecavanna — APPROVED** [review]
Looks good to me! I am comfortable with this being merged into the `main` branch. Once it's merged in, it'll be easier for me to test "all Berkeley migrators" in series.

**Comment by eecavanna** [comment]
Hi @turbomam, once you merge this branch into `main`, will you also publish a new prerelease package to PyPI? I think publishing a new package to PyPI will unblock https://github.com/microbiomedata/nmdc-runtime/issues/519#issuecomment-2116597460 (I'm getting some path-related errors when installing the package directly from the GitHub repo as opposed to from PyPI).

**Review by turbomam — APPROVED** [review]

**Pull request merged**
